From 985adba60caebfd8e58fd0c67f1f60fe78211518 Mon Sep 17 00:00:00 2001
From: Pillar Zuo <zuobaozhu@gmail.com>
Date: Wed, 7 Sep 2016 19:55:11 +0800
Subject: [PATCH 7/7] update ralink_gdma.c

Signed-off-by: Pillar Zuo <zuobaozhu@gmail.com>
---
 sound/soc/mtk/ralink_gdma.c | 1 -
 1 file changed, 1 deletion(-)

diff --git a/sound/soc/mtk/ralink_gdma.c b/sound/soc/mtk/ralink_gdma.c
index e072296..fda3932 100644
--- a/sound/soc/mtk/ralink_gdma.c
+++ b/sound/soc/mtk/ralink_gdma.c
@@ -806,7 +806,6 @@ irqreturn_t GdmaIrqHandler(
 	    }
 	}
      }	
-	//mdelay(15);
     //write 1 clear
 #if defined (CONFIG_RALINK_RT3052)	
      GDMA_WRITE_REG(RALINK_GDMAISTS, GdmaDoneStatus); 
-- 
1.9.1

